### ISSUE TITLE
Snakemake API

### ISSUE THREAD

**Issue by gilmorera** [snakemake/snakemake#2908]
Hi @johanneskoester et. al.  I've really been enjoying the new dataclass approach to the Python API for snakemake.  I've been successfully running some basic versions of my pipeline locally on my Mac, and I'm about to transition to an HPC environment.  My only holdup now is figuring out the 'executor_settings' that are defined in the dagAPI.

I know there isn't a lot of support for this but I have a few strategies that I'm thinking about implementing to make it work.  

1.  Import executor settings from one of the plugins.  In my case it would be the cluster-sync plugin executlor settings and in the near future the slurm plugin's executor settings.

2.  This is much less desirable, but I thought I could also use a configfile that applies those executor plugin settings.

**Comment by gilmorera** [comment]
Just looping back around to this question:
> Do you have any advice for using the general API along with the API from one of the executor plugins?

😃

**Comment by standage** [comment]
Glad it's not just me!

**Comment by cmeesters** [comment]
To be honest: I do not understand the question. Perhaps, if either of you, can elaborate your aim?

The SLURM executor plugin has no "API" designed to be imported by a user (it is, of course, possible to do just that, if the plugin is in your environment - if this is your starting point, I would be curious to learn what you want to achieve). Then again, your goal might be something entirely different.

**Comment by gilmorera** [comment]
Well there are two parts to the question..

1. using the official snakemake python modules to run a workflow - https://snakemake-api.readthedocs.io/en/stable/api_reference/snakemake_api.html
2. using the executor plugins along with the python API..

The question is how do you do it?  The documentation is there, but very lacking for the first point.  And non-existing for the second point?

**Comment by gilmorera** [comment]
As a workaround, I've essentially reverse engineered the main function in the cli module:  https://github.com/snakemake/snakemake/blob/19a2ab76b475c5d7728624618bec75a7f1411e8b/snakemake/cli.py#L2168-L2177

```python
 def run_snakemake_command(self, sys_exit=True):
        """This method is used to run snakemake cli commands using python.  It is a wrapper
        around the args_to_api and get_argument_parser functions in the snakemake.cli module, which are used to parse
        the command line arguments into various snakemake dataclasses (e.g. snakemake.settings).
        """

        try:
            logging.setup_logger()
            parser = self.argument_parser
            args = self.set_cli_args()
            success = args_to_api(args, parser)
        except Exception as e:
            print_exception(e)
            if sys_exit:
                sys.exit(1)
        if sys_exit:
            sys.exit(0 if success else 1)
```

**Comment by gilmorera** [comment]
And actually if you're interested @standage @cmeesters I can probably publish my code in a public repository pending approval.  It's pretty simple for the most part.

**Comment by standage** [comment]
> Perhaps, if either of you, can elaborate your aim?

My aim, as articulated in #2792, is to understand how to launch a Snakemake workflow from Python. I was able to do the following with Snakemake version <= 7, but the Snakemake version 8 documentation doesn't make it clear how to do this or something like it.

```python
from snakemake import snakemake
snakemake("workflow.smk", config=cfg)
```

**Comment by johanneskoester** [comment]
Indeed, the API docs don't really explain this yet. However, it should hopefully be pretty straightforward when looking at how the api is used here: https://github.com/snakemake/snakemake/blob/main/snakemake/cli.py#L1864

In short: you get the executor settings by loading the plugin by name and retrieving the settings:
```python
executor_plugin = ExecutorPluginRegistry().get_plugin("slurm")
# manually pass args to the respective class of the plugin:
executor_settings = executor_plugin.settings_cls(...)
# or infer from CLI args:
executor_settings = executor_plugin.get_settings(args)
```

Note that it indeed was a bit more complex before when one would have wanted to use one of the builtin executors like local or dryrun. I am working on fixing this here: https://github.com/snakemake/snakemake/pull/3114

@standage the basic idea of how snakemake is called now is described here:
https://snakemake-api.readthedocs.io/en/stable/api_reference/snakemake_api.html

**Comment by johanneskoester** [comment]
I keep this open in order to highlight the fact that we should extend the api docs in this regard.

**Comment by gilmorera** [comment]
@johanneskoester  thanks for the info.  And better API docs would certainly be helpful on my end 👍

**Comment by irr-github** [comment]
I have the related issue that the API example does not work for the recommended Snakemake8 project structure as the config file is searched for in project/Workflow/scripts and not project/configs by default.
This can be fixed passing the config settings explictly
`config_settings = sm_api.ConfigSettings(configfiles=cfg_paths)`
but it's not documented and the 'recommended project structure' should really be supported
